src/source/statsd.rs

```diff
@@ -106,55 +107,79 @@ fn handle_udp(
     mut chans: util::Channel,
     tags: &sync::Arc<metric::TagMap>,
     parse_config: &sync::Arc<StatsdParseConfig>,
-    socket: &UdpSocket,
+    conns: &util::TokenSlab<mio::net::UdpSocket>,
+    poll: &mio::Poll,
 ) {
     let mut buf = vec![0; 16_250];
     let mut metrics = Vec::new();
     let basic_metric = sync::Arc::new(Some(
         metric::Telemetry::default().overlay_tags_from_map(tags),
     ));
     loop {
-        let (len, _) = match socket.recv_from(&mut buf) {
-            Ok(r) => r,
-            Err(e) => panic!(format!("Could not read UDP socket with error {:?}", e)),
-        };
-        match str::from_utf8(&buf[..len]) {
-            Ok(val) => if parse_statsd(val, &mut metrics, &basic_metric, parse_config)
-            {
-                for m in metrics.drain(..) {
-                    send(&mut chans, metric::Event::new_telemetry(m));
+        let mut events = mio::Events::with_capacity(1024);
+        match poll.poll(&mut events, None) {
+            Ok(_num_events) => for event in events {
+                match event.token() {
+                    constants::SYSTEM => return,
+                    token => {
+                        // Get the socket to receive from:
+                        let socket = &conns[token];
+
+                        let (len, _) = match socket.recv_from(&mut buf) {
+                            Ok(r) => r,
+                            Err(e) => panic!(format!(
```

Comment:
In the event that we can't read from the socket here, what ought we to do? Is this a catastrophic failure – as indicated by the panic? – or something that can be recovered from? I'm not sure of the answer.

Reply:
It's what we did before: https://github.com/postmates/cernan/pull/339/files/61641dd333ad7f34e683ea2c86e42c68adf16612#diff-5bfbf0befe1a07571883accc30ffb6e7L119